### DISCUSSION TITLE
aligning text colors with background colors for better contrast

### DISCUSSION THREAD

**Discussion by eli-kha** [zauberzeug/nicegui#986] [Ideas / Feature Requests]
### Question

In the Quasar documentation page for the color pallet ([link](https://quasar.dev/style/color-palette)) you can see that the text color changes to create a good contrast.
When you set colors via `ui.colors` or directly with the elements the text color does not adapt.
It seems to be white by default and black if you use tailwinds colors.

```python
from nicegui import ui

# white despite bad contrast
ui.button('quasar orange-1', color='orange-1')
# white with good contrast
ui.button('quasar orange-10', color='orange-10')
# black with good contrast
ui.button('tailwind orange-100', color='orange-100')
# black despite bad contrast
ui.button('tailwind orange-900', color='orange-900')

ui.run(native=True)
```

Is there a way to autoselect the text color when using the pallet colors via classes, props or ui.color? I couldn't find any clear method of doing this
`colors.py` could be adapted to handle this but it wouldn't deal with the global theme if I understood correctly.

**Comment by rodja** [comment]
Maybe I'm misunderstanding something, but for me the Quasar color pallet is not adapting the text color. It's white even for the very bright tones:

<img width="776" alt="Screen Shot 2023-06-04 at 08 39 14" src="https://github.com/zauberzeug/nicegui/assets/131391/9782d7b9-e14a-4e1d-8dd1-63cdb356b57c">

**Reply by eli-kha** [reply]
This is because of your dark theme. Switch to light.
![image](https://github.com/zauberzeug/nicegui/assets/20465057/c9848732-1001-443d-bc6e-c5d766f85b3b)
![image](https://github.com/zauberzeug/nicegui/assets/20465057/ceef36a8-d613-4baf-9ea1-b9ec57b2442f)

Note that when you use white text with `red-2`, which is not the worst case, you're failing accessibility tests:
![image](https://github.com/zauberzeug/nicegui/assets/20465057/0826b74a-f045-413f-b2b4-4cebc8be77ae)

Whereas with a black font:
![image](https://github.com/zauberzeug/nicegui/assets/20465057/932e165d-4ca6-4d37-8d82-d3446f1b4a78)

*tested with [DevToys](https://devtoys.app/)
Side note: maybe there's an issue with the Quasar site itself? or the dark theme feature?

**Comment by rodja** [comment]
I think the color palette in the Quasar documentation is doing the foreground selection "by hand":

https://github.com/quasarframework/quasar/blob/324e8b740e1796b721124dcc7a9bcf1dfa137430/docs/src/pages/style/color-palette/ColorList.vue#L15

They simply pick the white text for larger colors and use the default otherwise. In dark-mode the default is white.

Doing auto-magic foreground color selection directly in NiceGUI core would be super cool but very hard to get right for all use-cases. I suggest you try to implement something which works for you by deriving from ui.button in your project and then share the result here.

**Reply by eli-kha** [reply]
I implemented based on W3 guidelines.
Have a look at the gist: https://gist.github.com/eli-kha/4b958a888b64230cc2b6eb9351f23750

**Reply by eli-kha** [reply]
Note that I only handled the HEX code case.
For quasar we can use [getpalettecolor](https://quasar.dev/quasar-utils/color-utils#helper-getpalettecolor)
For tailwind, I'm not sure but there's probably a way.

**Reply by rodja** [reply]
Thanks for the gist. Because Quasar and Tailwind colors are already ordered by hue, you could simply check the color number to decide which foreground color should be used instead of computing the hex. Similar as Quasar does it in their color palette documentation.
I think it would be cool to integrate your `set_text_color` into `ui.button` when the color parameter is provided. That way things are "nice" for the straight forward cases. And when users want to have full control over the colors they can always use the `.classes` or `.style` methods.
Would you like to create a pull request for this? That way you will be part of the NiceGUI contributors list.  See [CONTRIBUTING.md](https://github.com/zauberzeug/nicegui/blob/main/CONTRIBUTING.md#pull-requests) for a short explanation.

**Reply by eli-kha** [reply]
I can do the PR but I want to clear up a few questions first.

- I think that we do need to read the RGB colors:
  - In tailwinds the tipping point changes between specific colors in the pallet
  - In Quasar the hue change is not linear + we need access to the brand colors
  - I'll see how I can make this seamless as much as possible
- What about other elements that use the theme colors? like ui.date and others with a background? I'd like to solve this for everything

**Reply by falkoschindler** [reply]
@eli-kha I just refactored the color implementation and introduced two color mixins `TextColorElement` and `BackgroundColorElement`: [8f8aaef](https://github.com/zauberzeug/nicegui/commit/8f8aaef7c784f99e27282d44a3c3cf1d0b3e36c8). This is something I had in mind for quite a while and which will probably simplify any further improvements related to element colors.

My idea would be to add the text-color generator to `BackgroundColorElement.__init__`: If `isinstance(self, TextColorElement)`, we can call a method from `TextColorElement` to automatically set an adequate text color.

Further thoughts on this matter:
- [getpalettecolor](https://quasar.dev/quasar-utils/color-utils#helper-getpalettecolor) might be impractical since we would need to run JavaScript code on the client. At construction time there might not be any client connected. So maybe a hardcoded lookup table might be the easier way to go for Tailwind and Quasar color palettes.
- We should preserve the possibility to pick a custom text color. So the automatic text color should only be applied if no text color is specified. This is a bit tricky, because the color mixins don't have access to the default color values of individual elements. So maybe the elements like `ui.avatar` should only pass a `text_color` to `TextColorElement` if it differs from the default value - and `None` otherwise. This way the `TextColorElement` can decide whether to auto-generate a text color or not, once the auto-generation is triggered by the `BackgroundColorElement`.
